src/Bicep.Core/SourceCode/SourceArchive.cs

```diff
@@ -115,53 +130,81 @@ private record ArchiveMetadata(
         );
 
+        // A single SourceFiles entry in the metadata.json file
         private partial record SourceFileInfoEntry(
             // IF ADDING TO THIS: Remember both forwards and backwards compatibility.
             // E.g., previous versions must be able to deal with unrecognized source kinds.
             // (but see CurrentMetadataVersion for breaking changes)
             string Path,        // the location, relative to the main.bicep file's folder, for the file that will be shown to the end user (required in all Bicep versions)
             string ArchivePath, // the location (relative to root) of where the file is stored in the archive
-            string Kind         // kind of source
+            string Kind,        // kind of source (SourceKind)
+            string? SourceArtifactId = null // Points to an external artifact that contains the source for this module (e.g. "br:contoso.io/test/module1:v1"), appears in v0.26 and higher
         );
 
         #endregion
 
         public static ResultWithException<SourceArchive> UnpackFromStream(Stream stream)
         {
-            var archive = new SourceArchive(stream);
-            if (archive.GetRequiredBicepVersionMessage() is string message)
+            try
             {
-                return new(new Exception(message));
+                var archive = new SourceArchive(stream);
+                if (archive.GetRequiredBicepVersionMessage() is string message)
+                {
+                    return new(new Exception(message));
+                }
+                else
+                {
+                    return new(archive);
+                }
             }
-            else
+            catch (Exception ex)
             {
-                return new(archive);
+                return new(ex);
             }
         }
 
         /// <summary>
         /// Bundles all the sources from a compilation group (thus source for a bicep file and all its dependencies
         /// in JSON form) into an archive (as a stream)
         /// </summary>
-        /// <returns>A .tar.gz file as a binary stream</returns>
-        public static Stream PackSourcesIntoStream(SourceFileGrouping sourceFileGrouping, string? cacheRoot)
+        /// <returns>A .tgz file as a binary stream</returns>
+        public static Stream PackSourcesIntoStream(IModuleDispatcher moduleDispatcher, SourceFileGrouping sourceFileGrouping, string? cacheRoot)
         {
+            // Find the artifact reference for each source file of an external module that was published with sources
+            var uriToArtifactReference = sourceFileGrouping.FileUriResultByBicepSourceFileByArtifactReferenceSyntax
+                .SelectMany(outerKvp => outerKvp.Value, (outerKvp, innerKvp) => (bicep: outerKvp.Key, syntax: innerKvp.Key, uri: innerKvp.Value.TryUnwrap()))
```

Comment:
Here you may be packing providers alongside with the sources. You should probably filter by type. So that you are only packing Module objects

Reply:
Yep, forgot about that, thanks.

Reply:
Had it in the original code.